### DISCUSSION TITLE
Невозможно положить словарь в Property

### DISCUSSION THREAD

**Discussion by forwork10** [Sokol-Games-GmbH/PixelCombats.PublicApi#110] [Q&A] [answered]
Никак не получается положить словарь в проперти, всё это не будет работать:
```javascript
var prop=Properties.GetContext().Get("prop");

prop.Value={"a":1};
i=prop.Value["a"];

prop.Value=[{"a":1}];
i=prop.Value[0]["a"];
```
Возможно просто недоработка...

**Comment by kkohno** [comment, marked as answer]
Свойства поддерживают ограниченные типы данных. Хранят только примитивы: bool, short, int, long, float, string, 
Попробуйте сериализовать в строку и сохранить в свойство, потом когда читаете  придется обратно десериализовать.

**Comment by forwork10** [comment]
это можно, но слишком затратно в плане производительности, при большом размере данных

**Comment by kkohno** [comment]
> это можно, но слишком затратно в плане производительности, при большом размере данных

Пока в режимах не рассчитано на большие данные. Это все-таки трафик всех игроков и его нужно экономить. Можете описать что хотите сделать?

**Comment by forwork10** [comment]
Сделать адекватное сохранение данных игрока например, где ключ - это айди игрока, а значение - любые данные для сохранения

**Comment by kkohno** [comment]
> Сделать адекватное сохранение данных игрока например, где ключ - это айди игрока, а значение - любые данные для сохранения

А например, что именно сохранить?  Просто для таких целей скорее всего сделаем специальные АПИ разных типов: 
* который будет сохранять данные, полученные в режиме на устройство игрока (это для сохранения опций больше подойдет)
* который будет сохранять данные в базу данных сервера - это уже для постоянного сохранения данных игроков
* в теории, но не факт, что это будет целесообразно, можно сделать еще спец АПИ для сохранения больших данных на одну катку... но там будут ограничения также по типам итд. потому, что там все будет упаковываться, чтобы не жрать ресурсы

**Comment by Kukushechka7-ezx** [comment]
Что затратно?.. берешь и json кладёшь туда

**Comment by forwork10** [comment]
опять ты свои 5 копеек вставил... интерпретация строки в json очень затратна если ты не знал...

**Comment by Kukushechka7-ezx** [comment]
Нечего не затратно,лол, сам проверь

**Comment by kkohno** [comment]
> опять ты свои 5 копеек вставил... интерпретация строки в json очень затратна если ты не знал...

ну оно затратно конечно, просто это если делать редко то нормально. Например когда загружается игра. Главное не херачить в цикле))

**Comment by Kukushechka7-ezx** [comment]
Он про цикл то и имел наверное) то же сохранение при выходе не будет затратно

**Comment by forwork10** [comment]
будет, оно обрабатывается на всех клиентах

**Comment by Kukushechka7-ezx** [comment]
...

**Comment by kkohno** [comment]
в таймере тоже конечно лучше избегать... еще смотря сколько там данных лежит. если очень много то вообще фриз у всех будет

**Comment by Kukushechka7-ezx** [comment]
Ну таймер тут и не нужен

**Comment by kkohno** [comment]
предлагаю тему в дискуссию перенести. тут нет бага, но вот как возможное улучшение - тут то, что я писал выше точно делать будем

**Comment by justqstn** [comment]
возможное решение есть
использовать массив ключей и массив со значениями
```let keys = ["Kills", "Deaths"];
let default_values = [10, 10];
function load_data(player)
{
  if (Properties.GetContext().Get(player.Id + "save").Value) {
    keys.forEach(function(prop, index) {
      player.Properties.Get(prop).Value = Properties.GetContext().Get(player.Id + "save").Value[index];
    });
  } else {
    keys.forEach(function(prop, index) {
      player.Properties.Get(prop).Value = default_values[index]; // для значений по дефолту
    });
  }
}

function save_data(player)
{
  let arr = [];
  keys.forEach(function(prop) {
    arr.push(player.Properties.Get(prop).Value);
  });
  Properties.GetContext().Get(player.Id + "save").Value = arr;
}

**Reply by justqstn** [reply]
@kkohno ответ :)

**Reply by kkohno** [reply]
ну, это уже обходные пути все , конечно, но можно, если работает. просто может иногда падать например, если много хранить
